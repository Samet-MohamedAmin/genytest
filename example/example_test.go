package example_test

import (
	"genytest/example"
	"io/ioutil"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	approvals "github.com/approvals/go-approval-tests"
	"gopkg.in/yaml.v3"
)

type testTemplate struct {
	Id     string `yaml:"id"`
	Input  int    `yaml:"input"`
	Output int    `Yaml:"output"`
}

const testCasesFile = "test_cases.yaml"

func readTestCases() []testTemplate {
	filename, _ := filepath.Abs(testCasesFile)
	yamlFile, err := ioutil.ReadFile(filename)

	if err != nil {
		panic(err)
	}

	var testCases []testTemplate

	err = yaml.Unmarshal(yamlFile, &testCases)
	if err != nil {
		panic(err)
	}

	return testCases
}

func funcName() string {
	pc, _, _, _ := runtime.Caller(1)
	nameFull := runtime.FuncForPC(pc).Name()
	nameEnd := filepath.Ext(nameFull)
	name := strings.TrimPrefix(nameEnd, ".")
	return name
}

func Transform(testCases any) string {
	tc := testCases.(testTemplate)

	tc.Output = example.Bla(tc.Input)

	output, err := yaml.Marshal([]testTemplate{tc})
	if err != nil {
		panic(err)
	}

	return string(output)
}

func TestBla(t *testing.T) {

	testCases := readTestCases()

	approvals.VerifyAll(t, funcName(), testCases, Transform)

}
