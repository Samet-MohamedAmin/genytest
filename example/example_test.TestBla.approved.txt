TestBla


- id: "0"
  input: 1
  output: 0

- id: "1"
  input: 4
  output: 0

- id: "2"
  input: 6
  output: 5

- id: "3"
  input: 5
  output: 5

- id: "4"
  input: 15
  output: 10

- id: "5"
  input: 9
  output: 5

- id: "6"
  input: 10
  output: 10
